src/microengine_utils/scanner.py

```diff
@@ -0,0 +1,213 @@
+    - Read the `ScanResult`'s fields to automatically figure out which metrics should be collected
+    - Merges `ScanResult` `metadata` with boilerplate scanner information from `EngineInfo`
+    """
+    def wrapper(scan_fn: 'Callable[[str, ArtifactType, bytes, Dict, str], ScanResult]') -> 'Callable':
+        async def scan_wrapper(
+            self,
+            guid: 'str',
+            artifact_type: 'ArtifactType',
+            content: 'bytes',
+            metadata: 'Dict',
+            chain: 'str',
+        ):
+            tags = [f'type:{ ArtifactType.to_string(artifact_type) }']  # e.g 'type:file' or 'type:url'
+
+            try:
+                if asyncio.iscoroutinefunction(scan_fn):
+                    start = perf_counter()
+                    scan = await scan_fn(guid, artifact_type, content, metadata, chain)
+                    statsd.timing(scan, perf_counter() - start)
+                else:
+                    with statsd.timer(SCAN_TIME):
+                        scan = scan_fn(guid, artifact_type, content, metadata, chain)
+
+                # invalid bit or verdict
+                if type(scan.verdict) is not bool or type(scan.bit) is not bool:
+                    statsd.increment(SCAN_TYPE_INVALID, tags=tags)
+
+                # successful scan, verdict reported
+                elif scan.bit:
+                    with suppress(AttributeError, KeyError):
+                        getter = itemgetter if isinstance(scan.metadata, Mapping) else attrgetter
+                        tags.append(f'malware_family:{getter("scan.metadata.malware_family")}')
+
+                    # malicious/benign metrics
+                    if verbose:
+                        statsd.increment(
+                            SCAN_VERDICT,
+                            tags=[*tags, 'verdict:malicious' if scan.verdict else 'verdict:benign']
+                        )
+
+                    statsd.increment(SCAN_SUCCESS, tags=tags)
+                # no result reported
+                elif not scan.bit:
+                    if verbose:
+                        statsd.increment(SCAN_NO_RESULT, tags=tags)
+
+            except BaseScanError as e:
+                # If we encountered a scan error, we still return a scan result with `scan_error`
+                # documenting the error which occurred
+                v = Verdict().set_malware_name('').set_extra('scan_error', e.event_name)
+                scan = ScanResult(bit=False, verdict=False, metadata=v)
+
+                # e.g 'scan_error:fileskippedscanerror'
+                tags.append(f'scan_error:{e.event_name}')
+                statsd.increment(SCAN_FAIL, tags=tags)
+            finally:
+                # If engines define `scanner_metadata`, we'll automatically include all of the
+                # boilerplate data (e.g `platform`, `machine`, `signature_info`, etc.)
+                if isinstance(getattr(self, 'scanner_metadata', None), EngineInfo):
```

Comment:
I'm not a fan of hardcoding a string for a variable the person has to add to their `Scanner`

We can discuss other options in the call.